Gems/Atom/RHI/DX12/Code/Source/Platform/Windows/RHI/Device_Windows.cpp

```diff
@@ -222,18 +222,312 @@ namespace AZ
+                        
+                        for (uint32_t contextIdx = 0; contextIdx < currentNode->BreadcrumbContextsCount; contextIdx++)
+                        {
+                            currentNode->pBreadcrumbContexts[contextIdx];
+                            AZ_TracePrintf("Device", "%d %S \n", currentNode->pBreadcrumbContexts[contextIdx].BreadcrumbIndex, currentNode->pBreadcrumbContexts[contextIdx].pContextString);
+                        }
+                        for (uint32_t index = 0; index < currentNode->BreadcrumbCount; index++)
+                        {
+                            if (hasError && index == completedBreadcrumbCount)
+                            {
+                                // where the error happened
+                                AZ_TracePrintf("Device", " ==========================Error start==================================\n");
+                            }
+                            AZ_TracePrintf("Device", "      %d %s\n", index, GetBreadcrumpOpString(currentNode->pCommandHistory[index]));
+                            if (hasError && index == completedBreadcrumbCount)
+                            {
+                                // where the error happened
+                                AZ_TracePrintf("Device", " ==========================Error end=============================\n");
+                            }
+                        }
+                        
+                        AZ_TracePrintf("Device", " ==================================================================\n");
+
+                        currentNode = currentNode->pNext;
+                    }
+                }
+
+                if (SUCCEEDED(pDred->GetPageFaultAllocationOutput1(&dredPageFaultOutput)))
+                {
+                }
+            }
+
+            AZ_TracePrintf("Device", " ===========================End of OnDeviceRemoved================================\n");
+
+            if (IsAftermathInitialized())
+            {
+                // Try outputting the name of the last scope that was executing on the GPU
+                // There is a good chance that is the cause of the GPU crash and should be investigated first
+                Aftermath::OutputLastScopeExecutingOnGPU(GetAftermathGPUCrashTracker());
+            }
+
+            SetDeviceRemoved();
+        }
+
+        void HandleDeviceRemoved(PVOID context, BOOLEAN)
+        {
+            Device* removedDevice = (Device*)context;
+            removedDevice->OnDeviceRemoved();
+        }
+
+        void Device::InitDeviceRemovalHandle()
+        {
+            // Create fence to detect device removal
+            Microsoft::WRL::ComPtr<ID3D12Fence> fencePtr;
+            if (FAILED(m_dx12Device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_GRAPHICS_PPV_ARGS(fencePtr.GetAddressOf()))))
+            {
+                return;
+            }
+            m_deviceFence = fencePtr.Get();
```

Comment:
Why does this need to be a member variable? I might be reading this incorrectly but it seems like this fence is a one-time use obejct.

Reply:
I think we need to keep the fence alive in case release it may cancel the SetEventOnCompletion()
At least from dx12's sample code, it's not using local variable. 
https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device5-removedevice#remarks